keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+```
+
+This does not work if type checking is strict. The ephemeralContainers field
+will be reported as unrecognized.
+
+This becomes far more compact if we offer scoping:
+
+```yaml
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      scopes: ["spec.containers[*]", "initContainers[*]", "spec.ephemeralContainers[*]"]
+```
+
+Note: This is the sort of policy where cluster managers would ideally be able to
+register the policy to validate ephemeralContainers _before_ upgrading to the
+version of Kubernetes where ephemeralContainers are available for use.
+
+Annotation to field migration example: https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/pod-security-policy/seccomp/template.yaml
+
+#### Use Case: Multiple policy definitions for different versions of CRD
+
+While version conversion allows for single policy definition. Cases for multiple
+policy definitions are:
+
+1. A policy author wishes to write a policy for both the v1 and v2 of a CRD
+because they wish to avoid incuring a CRD conversion webhook request, which
+would happen if they only offered a single policy (at either version).
+
+OR
+
+2. A policy author wishes to write a policy such that it can be evluated
+"shift-left" in a pre-submit check.
+
+Proposed solution:
+
+- Use a `matchPolicy: Exact` for the v1 policy.
+- use `matchPolicy: Equivalent` and an exclude match rules for `v1` for the
+  policy that handles v2+. This way if a v3 is added in the future, the policy
+  for v2 applies via version conversion by default.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy1.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   ["example.com"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["myCRD"]
+    matchPolicy: Exact
+  validations:
+  - rule: "object.v1fieldname == 'xyz'"
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy1.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   ["example.com"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["myCRD"]
+      effect: Exclude
+    - apiGroups:   ["example.com"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["myCRD"]
+      matchPolicy: Equivalent #is the default
+  validations:
+  - rule: "object.v2fieldname == 'xyz'"
+```
+
+#### Use Case: Migrating from validating webhook to validation policy
+
+Steps:
+
+1. Webhook is configured and in-use.
+2. `ValidatingAdmissionPolicy` created with `FailPolicy: Ignore`
+3. `ValidatingAdmissionPolicy` is monitored to ensure it behaves the same as te webhook (logs or audit annotations can be used)
+4. `ValidatingAdmissionPolicy` is updated to `FailPolicy: Fail`
+5. Webhook is configured with `FailPolicy: Ignore` (optional)
```

Comment:
you would want to be sure the webhook never returns an error.  If the admission policy is equivalent, then policy would always fail before webhooks are called.

Reply:
That works out nicely. Added.